File: ports/haskell/src/Unicode/Security/Display/SourceDisplayDivergence.hs
{-|
Module      : Unicode.Security.Display.SourceDisplayDivergence
Description : Aggregate "what a reviewer sees differs from what runs" detector.

Haskell port of @Unicode.Security.Display.SourceDisplayDivergence@ from
unicode-lean, transliterated byte-faithfully from the verified Rust reference
implementation (the display-layer aggregator, reason-code letter @D@).

Threat model. A single covert or identity trick may look individually benign,
but any hit means the rendered source diverges from its logical content; two or
more is a strong compound signal. This detector runs the five constituent
detectors on the same codepoint stream and aggregates: zero fire → 'Clear',
exactly one → pass-through that family's tag, two or more → 'Compound'.

What the detector reuses. This is pure aggregation over the port's own
constituent detectors — it introduces no predicate, table, or host library of
its own. It reuses, in this exact canonical order, the five core-family
detectors that already live in 'Unicode.Security.Policy':

  1. 'Unicode.Security.Policy.tagBlockFinding'          → 'TagBlock'
  2. 'Unicode.Security.Policy.variationSelectorFinding' → 'VariationSelector'
  3. 'Unicode.Security.Policy.zeroWidthFinding'         → 'ZeroWidth'
  4. 'Unicode.Security.Policy.bidiFinding'              → 'BidiControl'
  5. 'Unicode.Security.Policy.homoglyphFinding'         → 'IdentifierHomoglyph'

Each constituent's "is clear" accessor is the emptiness of the 'Finding' list it
returns: a non-empty list is a fire (the family's classification is non-clear).

No positions at this layer. By the Lean spec the per-family verdicts carry the
implicated positions; the aggregate verdict carries only the sub-threat tag, so
'Classification' holds no position list.
-}
module Unicode.Security.Display.SourceDisplayDivergence
  ( SubThreat
      ( TagBlock, VariationSelector, ZeroWidth, BidiControl, IdentifierHomoglyph
      , Compound
      )
  , subThreatTag
  , Classification (Clear, Hazard)
  , classificationIsClear
  , classificationTag
  , Verdict (Verdict, verdictInput, verdictClassify, verdictFires)
  , detect
  , reasonCode
  ) where

import Unicode.Security.Display.SourceDisplayAggregate
  ( Classification (Clear, Hazard)
  , SubThreat
      ( BidiControl, Compound, IdentifierHomoglyph, TagBlock, VariationSelector
      , ZeroWidth
      )
  , classificationIsClear
  , classificationTag
  , classify
  , firedFrom
  , reasonCode
  , subThreatTag
  )
import qualified Unicode.Security.Policy as Policy

-- ─────────────────────────────────────────────────────────────────────
-- §1 Types
-- ─────────────────────────────────────────────────────────────────────

-- | Verdict — the structured output of 'detect' (mirrors the Lean @Verdict@).
data Verdict = Verdict
  { verdictInput    :: ![Int]
    -- ^ The scanned input codepoints.
  , verdictClassify :: !Classification
    -- ^ The aggregate classification verdict.
  , verdictFires    :: ![SubThreat]
    -- ^ The constituent family tags that fired, in canonical order.
  }
  deriving stock (Eq, Show)

-- ─────────────────────────────────────────────────────────────────────
-- §2 Aggregation
-- ─────────────────────────────────────────────────────────────────────

-- | The five constituent detectors paired with the tag each contributes, in
-- canonical aggregation order. Each entry runs the port's own core-family
-- detector on the same input; a non-empty 'Finding' list is a fire.
constituents :: [Int] -> [[Policy.Finding]]
constituents input =
  [ Policy.tagBlockFinding input
  , Policy.variationSelectorFinding input
  , Policy.zeroWidthFinding input
  , Policy.bidiFinding input
  , Policy.homoglyphConstituentFinding input
  ]

-- | The constituent tags that fired on this input, in canonical order.
firedFamilies :: [Int] -> [SubThreat]
firedFamilies input = firedFrom (map (not . null) (constituents input))

-- | The SourceDisplayDivergence detection function. Runs the five constituent
-- detectors in canonical order and aggregates by how many fired, mirroring the
-- verified Rust reference exactly; see the module header for the constituent
-- inventory and tag mapping.
detect :: [Int] -> Verdict
detect input =
  Verdict
    { verdictInput    = input
    , verdictClassify = classify fires
    , verdictFires    = fires
    }
  where
    fires = firedFamilies input

File: ports/go/security/source_display_divergence.go
package security

// source-display-divergence — the aggregate "what a reviewer sees differs from
// what the machine runs" detector (the display-layer aggregator, layer D).
//
// Byte-faithful port of Unicode/Security/Display/SourceDisplayDivergence.lean,
// mirroring the verified Rust port
// security/display/source_display_divergence.rs.
//
// Threat model. Tier D1. A single covert or identity trick may be individually
// benign-looking, but any hit means the rendered source diverges from its
// logical content; two or more is a strong compound signal. This detector runs
// the port's own five constituent detectors on the same codepoint stream and
// aggregates: zero fire → clear, exactly one → pass-through that family's tag,
// two or more → Compound. Every constituent fires region-agnostically — payloads
// inside string literals or comments count.
//
// Constituents reused (this port's own, in canonical aggregation order):
//	1. tag-block-payload          isTagCharacter              → "TagBlock"
//	2. variation-selector-payload variationSelectorFinding    → "VariationSelector"
//	3. zero-width-payload         isZeroWidthPayload          → "ZeroWidth"
//	4. bidi-control-balance       isBidiEmbeddingControl      → "BidiControl"
//	5. homoglyph-confusable       homoglyphConfusableFinding  → "IdentifierHomoglyph"
//
// No new predicate, no new data file, no host library — pure aggregation over
// existing port code. Positions are empty at this layer by the Lean spec (the
// per-family verdicts carry them); this detector carries only the sub-threat.

// sddDetection is one source-display-divergence scan result. When fired is false
// the input is clear; a single constituent hit passes through its family tag,
// two or more yield "Compound". Positions are empty at this layer.
type sddDetection struct {
	sub   string
	fired bool
}

// isClear reports whether the detection is Clear (no constituent fired).
func (d sddDetection) isClear() bool {
	return !d.fired
}

// tag is the human-facing sub-threat tag; the second result is false when clear.
func (d sddDetection) tag() (string, bool) {
	if !d.fired {
		return "", false
	}
	return d.sub, true
}

// sddTagBlockFired reports whether the port's own tag-block-payload constituent
// classifies the input as non-clear. The family fires on any tag character, not
// only the ASCII-bearing span, so a LANGUAGE TAG or a lone CANCEL TAG counts
// here as it does in the family's own verdict.
func sddTagBlockFired(input []uint32) bool {
	return len(positionsWhere(input, isTagCharacter)) > 0
}

// sddVariationSelectorFired reports whether the port's own
// variation-selector-payload constituent classifies the input as non-clear.
func sddVariationSelectorFired(input []uint32) bool {
	_, ok := variationSelectorFinding(input)
	return ok
}

// sddZeroWidthFired reports whether the port's own zero-width-payload constituent
// classifies the input as non-clear. A ZWJ inside a registered emoji sequence
// and a ZWNJ in an RFC 5892 CONTEXTJ-valid position are present but carry
// meaning a reader depends on, so they do not make the constituent fire.
func sddZeroWidthFired(input []uint32) bool {
	positions := positionsWhere(input, isZeroWidthPayload)
	return len(positions) > 0 && hasSuspiciousZeroWidth(input, positions)
}

// sddBidiControlFired reads purpose, not presence (Lean: the C5 constituent is
// BidiControlPurpose.hasPurposelessControl). A Trojan Source payload balances
// its controls, so the balance verdict is blind to it; what every payload
// shares is a control span with nothing right-to-left to manage. A balanced
// embedding around a pure Arabic literal is purposeful and is not a divergence.
func sddBidiControlFired(input []uint32) bool {
	return hasPurposelessControl(input)
}

// sourceDisplayDivergenceDetect reads the input as the identifier its
// constituents assume. Mirrors the Lean detect, which is detectCore with the
// homoglyph family's identifier-reading verdict.
func sourceDisplayDivergenceDetect(input []uint32) sddDetection {
	_, homoglyphFired := homoglyphConfusableFindingCtx(input, homoglyphContext{})
	return sourceDisplayDivergenceDetectCore(input, homoglyphFired)
}

// sourceDisplayDivergenceDetectCore aggregates the five constituent detectors
// into a single D1 verdict with the homoglyph constituent supplied:
// homoglyphFired is whether the homoglyph family fired on this input under
// whatever reading the caller took (whole input, or per identifier token of
// running text). Mirrors the Lean detectCore, which the policy scan calls with
// the same verdict the homoglyph family reports. Constituents are tested in
// canonical aggregation order and each firing tag is collected; the verdict is
// then clear (0), a pass-through of the lone firing tag (1), or Compound (2 or
// more).
func sourceDisplayDivergenceDetectCore(input []uint32, homoglyphFired bool) sddDetection {
	fires := make([]string, 0, 5)
	if sddTagBlockFired(input) {
		fires = append(fires, "TagBlock")
	}
	if sddVariationSelectorFired(input) {
		fires = append(fires, "VariationSelector")
	}
	if sddZeroWidthFired(input) {
		fires = append(fires, "ZeroWidth")
	}
	if sddBidiControlFired(input) {
		fires = append(fires, "BidiControl")
	}
	if homoglyphFired {
		fires = append(fires, "IdentifierHomoglyph")
	}

	if len(fires) == 0 {
		return sddDetection{fired: false}
	}
	if len(fires) == 1 {
		return sddDetection{sub: fires[0], fired: true}
	}
	return sddDetection{sub: "Compound", fired: true}
}
